components/device-mgt/org.wso2.carbon.device.mgt.core/src/main/java/org/wso2/carbon/device/mgt/core/service/DeviceManagementProviderServiceImpl.java

```diff
@@ -716,6 +716,45 @@ public void unregisterDeviceManagementService(DeviceManagementService deviceMana
         }
     }
 
+
+    public List<Device> getDevicesByStatus(EnrolmentInfo.Status status) throws DeviceManagementException {
+        List<Device> devices = new ArrayList<Device>();
+        List<Device> allDevices;
+        try {
+            DeviceManagementDAOFactory.getConnection();
+            int tenantId = CarbonContext.getThreadLocalCarbonContext().getTenantId();
+            allDevices = deviceDAO.getDevicesByStatus(status, tenantId);
+
+        } catch (DeviceManagementDAOException e) {
+            String errorMsg = "Error occurred while fetching the list of devices that matches to status: '"
+                              + status + "'";
+            log.error(errorMsg, e);
+            throw new DeviceManagementException(errorMsg, e);
+        } finally {
+
+            try {
+                DeviceManagementDAOFactory.closeConnection();
+            } catch (DeviceManagementDAOException e) {
+                log.warn("Error occurred while closing the connection", e);
+            }
+        }
+
+        for (Device device : allDevices) {
+            Device dmsDevice =
+                    this.getPluginRepository().getDeviceManagementService(
+                            device.getType()).getDeviceManager().getDevice(
```

Comment:
You can use private method getDeviceManager() straightaway here to get an instance of corresponding device manager.